src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -2087,6 +2104,14 @@ export class SavedObjectsRepository {
     );
 
     const indexFound = statusCode !== 404;
+
+    // check if we have the elasticsearch header when doc.found is not true (false, undefined or null) and if we do, ensure it is Elasticsearch
+    const esServerSupported = isSupportedEsServer(headers);
```

Comment:
optional: we repeat the check in several places. It can be extracted in a helper.

Reply:
True, but other checks are also repeated in a few places. e.g. `indexNotFound`, `!isFoundGetResponse(body)` etc. 

Maybe I'm misunderstanding _what_ should be extracted, we've already extracted the functionality to ensure that the header is present in the response.

Reply:
That's what confuses me. Why in some places do we check `isFoundGetResponse` but in some `indexNotFound`? 
I'd expect we have the same check for all the methods.

Reply:
> Why in some places do we check isFoundGetResponse but in some indexNotFound?
I'd expect we have the same check for all the methods.

I don't know either. It might be worth trying to clean this up with an overhaul but that's a separate issue.

Reply:
I'll have to go through the git history for the methods to dig up the possible causes for getting `404` responses.

Reply:
Maybe @rudolf might shed light on the purpose of different conditions here?

Reply:
I can't think of any reason to have different implementations between the methods. When an index isn't found or the document isn't found we should return a 404.

The index not found behaviour is mostly an edge case that I doubt ever happens in practise and when it does there's much bigger problems. In the future I want us to return a 500, when the index isn't there Kibana is broken and in an unrecoverable state.

Reply:
> When an index isn't found or the document isn't found we should return a 404

I've tried to keep the scope pretty small in this work and only add the header check on the response to `get`, `update` and `delete` when we would otherwise be throwing a 404.

Reply:
> That's what confuses me. Why in some places do we check `isFoundGetResponse` but in some `indexNotFound`?
> I'd expect we have the same check for all the methods.

I believe the reason is that the `update` and `delete` APIs don't return the same `doc.found` property, though they do return a `result` property that may be either `updated` or `deleted` or `not_found`.

I don't think we should change this as part of this PR, it feels out of scope to me.

Reply:
I've created https://github.com/elastic/kibana/issues/107944 from this comment thread to make sure we don't lose sight of the discussion.